base/broadcast.jl

```diff
@@ -1153,15 +1153,15 @@ broadcasted(::DefaultArrayStyle{1}, ::typeof(-), r::LinRange, x::Number) = LinRa
 broadcasted(::DefaultArrayStyle{1}, ::typeof(-), x::Number, r::LinRange) = LinRange(x - r.start, x - r.stop, length(r))
 broadcasted(::DefaultArrayStyle{1}, ::typeof(-), r1::AbstractRange, r2::AbstractRange) = r1 - r2
 
-broadcasted(::DefaultArrayStyle{1}, ::typeof(*), x::Number, r::AbstractRange) = range(x*first(r), step=x*step(r), length=length(r))
-broadcasted(::DefaultArrayStyle{1}, ::typeof(*), x::Number, r::StepRangeLen{T}) where {T} =
-    StepRangeLen{typeof(x*T(r.ref))}(x*r.ref, x*r.step, length(r), r.offset)
+broadcasted(::DefaultArrayStyle{1}, ::typeof(*), x::Number, r::AbstractRange) = StepRangeLen(x*first(r), x*step(r), length(r))
+broadcasted(::DefaultArrayStyle{1}, ::typeof(*), x::Number, r::StepRangeLen{T}) where {T} = StepRangeLen{typeof(x*T(r.ref))}(x*r.ref, x*r.step, length(r), r.offset)
 broadcasted(::DefaultArrayStyle{1}, ::typeof(*), x::Number, r::LinRange) = LinRange(x * r.start, x * r.stop, r.len)
-# separate in case of noncommutative multiplication
-broadcasted(::DefaultArrayStyle{1}, ::typeof(*), r::AbstractRange, x::Number) = range(first(r)*x, step=step(r)*x, length=length(r))
-broadcasted(::DefaultArrayStyle{1}, ::typeof(*), r::StepRangeLen{T}, x::Number) where {T} =
-    StepRangeLen{typeof(T(r.ref)*x)}(r.ref*x, r.step*x, length(r), r.offset)
+broadcasted(::DefaultArrayStyle{1}, ::typeof(*), x::AbstractFloat, r::OrdinalRange{<:Integer}) = Base.range_start_step_length(x*first(r), x, length(r))
+# separate in case of noncommutative multiplication:
+broadcasted(::DefaultArrayStyle{1}, ::typeof(*), r::AbstractRange, x::Number) = StepRangeLen(first(r)*x, step(r)*x, length(r))
+broadcasted(::DefaultArrayStyle{1}, ::typeof(*), r::StepRangeLen{T}, x::Number) where {T} = StepRangeLen{typeof(T(r.ref)*x)}(r.ref*x, r.step*x, length(r), r.offset)
 broadcasted(::DefaultArrayStyle{1}, ::typeof(*), r::LinRange, x::Number) = LinRange(r.start * x, r.stop * x, r.len)
+broadcasted(::DefaultArrayStyle{1}, ::typeof(*), r::OrdinalRange{<:Integer}, x::AbstractFloat) = Base.range_start_step_length(first(r)*x, x, length(r))
```

Comment:
The last method here handles the case needed for AbstractFFTs tests, above, of float * UnitRange -> range with twiceprecision. 

And `(-2:2) * 0.0` works because for floats `range_start_step_length` does sometimes allow zero step:
```
julia> Base.range_start_step_length(1.0, 0, 5)
StepRangeLen(1.0, 0.0, 5)

julia> Base.range_start_step_length(1, 0, 5)  # goes to 1st method not 4th
ERROR: ArgumentError: step cannot be zero
```
But probably it could just be `OrdinalRange`. Are there other weird cases we should think about?

Reply:
I wonder whether `Base.range_start_step_length(1, 0, 5)` should just return a `StepRangeLen`. Do you expect that to be too breaking?

Reply:
I haven't thought about it. Was hoping to change as little as possible, focused on calculations that may produce zero step, rather than constructors. 

This one is called by `range(1, step=0, length=10)` which would otherwise make an integer StepRange. Whereas `range(1, 1, length=10)` works, and makes a floating point range. It is a bit weird that some constructors work and some don't, but this isn't changed here.

Reply:
Agree, it's probably better not to change this behavior in this PR. Could you just add a short comment here explaining why we can't use `range_start_step_length`?

Reply:
Sure, see what you think of https://github.com/JuliaLang/julia/pull/40320/commits/56d9c27b617c5c9bb4b2fa720ffa3d9b0434e5a9

Reply:
Looks good! Let's run PkgEval one more time once tests pass, but otherwise this should be good to go.